screeps-game-api/src/objects/impls/room.rs

```diff
@@ -432,36 +436,148 @@ pub enum Path {
+            where
+                V: MapAccess<'de>,
+            {
+                let mut event_type = None;
+                let mut obj_id = None;
+                let mut data = None;
+                let mut data_buffer: Option<serde_json::Value> = None;
+
+                while let Some(key) = map.next_key()? {
+                    match key {
+                        Field::Event => {
+                            if event_type.is_some() {
+                                return Err(de::Error::duplicate_field("event"));
+                            }
+                            event_type = Some(map.next_value()?);
+                        }
+                        Field::ObjectId => {
+                            if obj_id.is_some() {
+                                return Err(de::Error::duplicate_field("objectId"));
+                            }
+                            obj_id = Some(map.next_value()?);
+                        }
+                        Field::Data => {
+                            if data.is_some() {
+                                return Err(de::Error::duplicate_field("data"));
+                            }
+
+                            match event_type {
+                                None => data_buffer = map.next_value()?,
+                                Some(event_id) => {
+                                    data = match event_id {
+                                        1 => Some(EventType::Attack(map.next_value()?)),
+                                        2 => Some(EventType::ObjectDestroyed(map.next_value()?)),
+                                        3 => Some(EventType::AttackController),
+                                        4 => Some(EventType::Build(map.next_value()?)),
+                                        5 => Some(EventType::Harvest(map.next_value()?)),
+                                        6 => Some(EventType::Heal(map.next_value()?)),
+                                        7 => Some(EventType::Repair(map.next_value()?)),
+                                        8 => Some(EventType::ReserveController(map.next_value()?)),
+                                        9 => Some(EventType::UpgradeController(map.next_value()?)),
+                                        10 => Some(EventType::Exit(map.next_value()?)),
+                                        _ => {
+                                            return Err(de::Error::custom(format!(
+                                                "Event Type Unrecognized: {}",
+                                                event_id
+                                            )))
+                                        }
+                                    };
+                                }
+                            };
+                        }
+                    }
+                }
+
+                if data_buffer.is_some() && event_type.is_some() && data.is_none() {
```

Comment:
I think we should be able to remove the redundancy between these `is_some` calls and the `unwrap()` / `if let` below if we directly match on the values here. What would you think of something like this?

```rust
if data.is_none() {
    if let (Some(data_buffer), Some(event_type)) = (data_buffer, event_type) {
        ...
    }
}
```

Reply:
I had avoided doing that because I didn't want to get 3 levels deep, but that fixes my concern nicely. Looks great!

Reply:
Cool!

We might even be able to do 
```rust
if let (Some(data_buffer), Some(event_type), None) = (data_buffer, event_type, data) {
```

but I'm not 100% sure rust will allow that since we're modifying `data` inside and lifetimes are lexical (can't wait for non-lexical lifetimes but that's a whole other topic).